architecture-docs/adr-002.md

```diff
@@ -0,0 +1,490 @@
+    /// `sum_average_stake` can be calculated using `ProposerTracker::get_sum_average_stake()`
+    #[inline]
+    pub fn calculate_reward(&self, sum_average_stake: Coin) -> Coin {
+        self.config.calculate_reward(total_stake, self.minted)
+    }
+
+    /// Decays `tau` based on `decay_factor` in rewards config
+    #[inline]
+    pub fn decay_tau(&mut self) {
+        self.config.decay_tau()
+    }
+
+    /// Updates rewards pool state with given values
+    pub fn update(
+        &mut self, 
+        period_bonus: Coin, 
+        last_block_height: BlockHeight, 
+        last_distribution_time: Timespec, 
+        minted: Coin,
+    ) {
+        self.period_bonus = period_bonus;
+        self.last_block_height = last_block_height;
+        self.last_distribution_time = last_distribution_time;
+        self.minted = minted;
+    }
+}
+```
+
+`ProposerTracker` tracks the number of blocks proposed by any validator and also maintains all the validator set changes
+in current reward period.
+
+```rust
+/// Tracks the number of blocks proposed by validators and also their voting power changes
+pub struct ProposerTracker {
+    /// Number of blocks a validator proposed in current reward period
+    proposer_count: BTreeMap<StakedStateAddress, u64>,
+    /// Changes in active validator set in current reward period
+    validator_set_changes: Vec<(Timespec, Vec<Validator>)>,
+}
+
+impl ProposerTracker {
+    /// Initializes a new rewards period
+    #[inline]
+    pub initialize(&mut self, time: Timespec, active_validator_set: Vec<Validator>) {
+        self.clear();
+        self.initial_voting_powers = vec![(time, active_validator_set)];
+    }
+
+    /// Clears proposer history (should be called when rewards are distributed)
+    #[inline]
+    pub fn clear() {
+        self.proposer_count.clear();
+        self.validator_set_changes.clear();
+    }
+
+    /// Adds given staking address to proposed count
+    #[inline]
+    pub fn proposed(&mut self, staking_address: StakedStateAddress) {
+        self.proposer_count
+            .entry(staking_address)
+            .and_modify(|count| *count += 1)
+            .or_insert(1);
+    }
+
+    /// Updates the active validator set information for calculation of `sum_average_stake` used to calculate rewards.
+    #[inline]
+    pub fn validator_set_changed(&mut self, time: Timespec, active_validator_set: Vec<Validator>) {
+        self.validator_set_changes.push((time, active_validator_set))
+    }
+
+    /// Calculates the sum of average voting powers of each validator (takes voting power changes in account)
+    ///
+    /// # Example
+    ///
+    /// Let us assume that rewards period is 24 hours and there is only one active validator. For the first 23 hours,
+    /// their voting power was 50 and for the last hour, their voting power was 100. Then, the sum average stake should
+    /// be ((23 * 50) + (1 * 100)) / 24. This should be calculated for each validator and their sum should be returned.
```

Comment:
https://crypto-com.github.io/getting-started/staking.html#rewards

> S: total stakings at current block when reward distribution happens

not average

Reply:
I discussed this with @lezzokafka and we both agreed that we should average the staking power over rewards period.